src/external_acl.cc

```diff
@@ -951,58 +774,67 @@ static void
 externalAclHandleReply(void *data, const Helper::Reply &reply)
 {
     externalAclState *state = static_cast<externalAclState *>(data);
-    externalAclState *next;
-    ExternalACLEntryData entryData;
 
     debugs(82, 2, "reply=" << reply);
 
+    ExternalACLEntryPointer entry;
+    if (cbdataReferenceValid(state->def)) {
+        if (const auto ptr = state->def->cache.get(SBuf(state->key)))
+            entry = *ptr;
+    }
+
+    if (!entry)
+        entry = new ExternalACLEntry;
+
     if (reply.result == Helper::Okay)
-        entryData.result = ACCESS_ALLOWED;
+        entry->result = ACCESS_ALLOWED;
     else if (reply.result == Helper::Error)
-        entryData.result = ACCESS_DENIED;
+        entry->result = ACCESS_DENIED;
     else //BrokenHelper,TimedOut or Unknown. Should not cached.
-        entryData.result = ACCESS_DUNNO;
+        entry->result = ACCESS_DUNNO;
 
-    // XXX: make entryData store a proper Helper::Reply object instead of copying.
+    // XXX: make entry store a proper Helper::Reply object instead of copying.
 
-    entryData.notes.append(&reply.notes);
+    entry->notes.append(&reply.notes);
 
     const char *label = reply.notes.findFirst("tag");
     if (label != nullptr && *label != '\0')
-        entryData.tag = label;
+        entry->tag = label;
 
     label = reply.notes.findFirst("message");
     if (label != nullptr && *label != '\0')
-        entryData.message = label;
+        entry->message = label;
 
     label = reply.notes.findFirst("log");
     if (label != nullptr && *label != '\0')
-        entryData.log = label;
+        entry->log = label;
 
 #if USE_AUTH
     label = reply.notes.findFirst("user");
     if (label != nullptr && *label != '\0')
-        entryData.user = label;
+        entry->user = label;
 
     label = reply.notes.findFirst("password");
     if (label != nullptr && *label != '\0')
-        entryData.password = label;
+        entry->password = label;
 #endif
 
     // XXX: This state->def access conflicts with the cbdata validity check
     // below.
     dlinkDelete(&state->list, &state->def->queue);
 
-    ExternalACLEntryPointer entry;
-    if (cbdataReferenceValid(state->def))
-        entry = external_acl_cache_add(state->def, state->key, entryData);
```

Comment:
AFAICT, official code unconditinoally creates ExternalACLEntry (that stores helper lookup info) and passes it to ACLExternal::aclMatchExternal() via ACLFilledChecklist::extacl_entry hack. This approach is ugly as hell, but it does not rely on the cache to keep/pass fresh lookup information: Despite its name, external_acl_cache_add() may actually add nothing to the cache.

PR code removes ACLFilledChecklist::extacl_entry hack and relies on the cache to keep/pass fresh lookup information. This approach is conceptually wrong: Caching is an optimization that we cannot rely on to pass information. The code has to be written so that it continues to work correctly when caching is disabled or otherwise impossible.

Before refactoring to fix it, please create a test case that will fail (with the current PR code) when external ACL caching is disabled.